src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceGenerator.cs

```diff
@@ -0,0 +1,200 @@
+
+            if (syntaxXml == null)
+            {
+                context.ReportDiagnostic(Diagnostic.Create(s_MissingSyntaxXml, location: null));
+                return;
+            }
+
+            var syntaxXmlText = syntaxXml.GetText();
+
+            if (syntaxXmlText == null)
+            {
+                context.ReportDiagnostic(Diagnostic.Create(s_UnableToReadSyntaxXml, location: null));
+                return;
+            }
+
+            Tree tree;
+            var reader = XmlReader.Create(new SourceTextReader(syntaxXmlText), new XmlReaderSettings { DtdProcessing = DtdProcessing.Prohibit });
+
+            try
+            {
+                var serializer = new XmlSerializer(typeof(Tree));
+                tree = (Tree)serializer.Deserialize(reader);
+            }
+            catch (InvalidOperationException ex) when (ex.InnerException is XmlException)
+            {
+                var xmlException = (XmlException)ex.InnerException;
+
+                var line = syntaxXmlText.Lines[xmlException.LineNumber - 1]; // LineNumber is one-based.
+                int offset = xmlException.LinePosition - 1; // LinePosition is one-based
+                var position = line.Start + offset;
+                var span = new TextSpan(position, 0);
+                var lineSpan = syntaxXmlText.Lines.GetLinePositionSpan(span);
+
+                context.ReportDiagnostic(
+                    Diagnostic.Create(
+                        s_SyntaxXmlError,
+                        location: Location.Create(syntaxXml.Path, span, lineSpan),
+                        xmlException.Message));
+
+                return;
+            }
+
+            TreeFlattening.FlattenChildren(tree);
+
+            AddResult(writer => SourceWriter.WriteMain(writer, tree, context.CancellationToken), "Syntax.xml.Main.Generated.cs");
+            AddResult(writer => SourceWriter.WriteInternal(writer, tree, context.CancellationToken), "Syntax.xml.Internal.Generated.cs");
+            AddResult(writer => SourceWriter.WriteSyntax(writer, tree, context.CancellationToken), "Syntax.xml.Syntax.Generated.cs");
+
+            void AddResult(Action<TextWriter> writeFunction, string hintName)
+            {
+                // Write out the contents to a StringBuilder to avoid creating a single large string
+                // in memory
+                var stringBuilder = new StringBuilder();
+                using (var textWriter = new StringWriter(stringBuilder))
+                {
+                    writeFunction(textWriter);
+                }
+
+                // And create a SourceText from the StringBuilder, once again avoiding allocating a single massive string
+                context.AddSource(hintName, SourceText.From(new StringBuilderReader(stringBuilder), stringBuilder.Length, encoding: Encoding.UTF8));
+            }
+        }
+
+        private sealed class SourceTextReader : TextReader
```

Comment:
Creating a `TextReader` from `SourceText` seems like this is something that should be provided by Roslyn as well. Otherwise, people are more likely to create a `string` out of it than make their own reader, which is less efficient.